### PR TITLE
Don't modify the original SEXP with r_shape_to_buffer_adaptor().

### LINKED ISSUE #86: r_shape_to_buffer_adaptor() modifies the original SEXP
I have a few test cases for `rray`'s reducers where I accept an integer vector (not a 1D array of integers) and run something like `rray_sum()` on it which calls the `xt::sum()` reducer. I happened to notice that when I did this, my original integer vector was converted to an array (modified in place). After some sleuthing, I discovered that `r_shape_to_buffer_adaptor()` sets the `R_DimSymbol` on the original object when it is `NULL` (as it would be for a vector). 

It happens in these two places:
https://github.com/QuantStack/xtensor-r/blob/master/include/xtensor-r/rcontainer.hpp#L78
https://github.com/QuantStack/xtensor-r/blob/master/include/xtensor-r/rcontainer.hpp#L96

It looks something like:

``` r
library(rray)
x_int <- 1:10
x_int
#>  [1]  1  2  3  4  5  6  7  8  9 10
attributes(x_int)
#> NULL
class(x_int)
#> [1] "integer"
rray:::rray_reducer_cpp("sum", x_int, 0)
#> 55
x_int
#>  [1]  1  2  3  4  5  6  7  8  9 10
attributes(x_int)
#> $dim
#> [1] 10
class(x_int)
#> [1] "array"
```

It would be great if this didn't happen, and I don't think it needs to.

Below is a suggested modification for `r_shape_to_buffer_adaptor()` that gets all of the shape info without setting the `R_DimSymbol`. Things seem to work fine with this, and the original object isn't modified in place.

```cpp
        inline xbuffer_adaptor<int*> r_shape_to_buffer_adaptor(SEXP exp)
        {
            SEXP shape_sexp;
            SEXP dim = Rf_getAttrib(exp, R_DimSymbol);
            
            if (Rf_isNull(dim))
            {
                // This is a 0D scalar
                if (Rf_xlength(exp) == 1)
                {
                    return xbuffer_adaptor<int*>(nullptr, 0);
                }

                shape_sexp = Rcpp::IntegerVector::create(Rf_length(exp));
            }
            else 
            {
                shape_sexp = dim;
            }
            
            std::size_t n = (std::size_t)Rf_xlength(shape_sexp);
            return xbuffer_adaptor<int*>(
                Rcpp::internal::r_vector_start<INTSXP>(shape_sexp), n);
        }

        inline xbuffer_adaptor<int*> r_shape_to_buffer_adaptor(SEXP exp, std::size_t n)
        {
            SEXP shape_sexp;
            SEXP dim = Rf_getAttrib(exp, R_DimSymbol);
            
            if (n == 0)
            {
                return xbuffer_adaptor<int*>(nullptr, 0);
            }
            
            if (Rf_isNull(dim))
            {
                shape_sexp = Rcpp::IntegerVector::create(Rf_length(exp));
            }
            else {
                shape_sexp = dim;
            }

            if (n != (std::size_t)Rf_xlength(shape_sexp))
            {
                throw std::runtime_error("Could not convert shape for rtensor. Dimensions don't match.");
            }
            return xbuffer_adaptor<int*>(
                Rcpp::internal::r_vector_start<INTSXP>(shape_sexp), n);
        }
```

### PR DESCRIPTION
**Pull request by DavisVaughan**
Closes #86

### CODE REVIEW

File: include/xtensor-r/rcontainer.hpp [hunk 1]
```diff
@@ -66,37 +66,47 @@ namespace xt
...
             }
-            if (Rf_isNull(Rf_getAttrib(exp, R_DimSymbol)))
+
+            if (Rf_isNull(dim))
             {
-                auto d = Rcpp::IntegerVector::create(Rf_length(exp));
-                Rf_setAttrib(exp, R_DimSymbol, d);
+                shape_sexp = Rcpp::IntegerVector::create(Rf_length(exp));
+            }
+            else {
```
Comment (wolfv) [inline include/xtensor-r/rcontainer.hpp]: just one new-line here and we can merge! sorry for taking so long.

**Comment by wolfv** [comment]
Thanks for this PR!

File: include/xtensor-r/rcontainer.hpp (hunk 1 above)
Comment (DavisVaughan) [reply include/xtensor-r/rcontainer.hpp]: ✅

**Comment by DavisVaughan** [comment]
Done!

**Pull request merged**